### DISCUSSION TITLE
Auxvariable vs non-linear variable for chemical potential-driven diffusion coupling

### DISCUSSION THREAD

**Discussion by LukeFisherUKAEA** [idaholab/moose#20160] [Q&A General] [answered]
Hello,

I'm looking to implement a chemical potential-driven diffusion model, with the following governing equations:

(1)    dC/dt = ( (D*C) / (R*T) ) * d^2mu / dx^2

(2)    mu = mu_0 + R*T*ln ( C / C_max )

My understanding is that for equation (1) we can use an ADTimeDerivative with variable C, and an ADMatDiffusion with variable = mu coupled to v = C.  An ADParsedMaterial block calculates the coefficient.

I'm not completely clear whether variables linked to kernels must in all cases be non-linear, rather than auxiliary?  It seems that a 'nice' way of setting this problem up would be to set _mu_ as an auxvariable function, which is calculated on every timestep by equation (2) in a ParsedAux block.  Alternatively, _mu_ could be assigned as a non-linear variable and equation (2) evaluated as a material property and _mu_ calculated through a ADMatCoupledForce block.  This seems more difficult.

Which is the best way of doing this, and is it the case that auxvariables can never be linked to kernel calculations?  Thanks.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Well it really depends what you want your numerical / time integration scheme to be.
If you want to be implicit in C and mu, then use an implicit time scheme and make both non-linear variables. 
If you want to be implicit in C and explicit in mu, then use an implicit time scheme and make C a non-linear variable. Then you can choose whether mu should be computed at the beginning or the end of a time step (matters especially for first time step), based on the execute_on of the auxkernel that computes it.

By the way, ADMatDiffusion doesnt do what you think here, because setting the variable to mu does compute the right term, but it places it in the wrong equation.
You have to think of the variable parameter as two things:
- it's what goes in the formula for computing that term. 
- it's the name of equation (equation for C, equation for mu) that you add that term to (that term is part of the residual)

So for that kernel:
v = mu
variable = C
is more correct. You will still need to capture the C-dependency in the diffusivity material property.

Guillaume

**Reply by LukeFisherUKAEA** [reply]
Thank you for the advice and the pointer on setting variables to the correct identifier inside kernels - this has cleared up some confusion!  I've gone with the implicit C/explicit mu approach.  I'm finding that the use of a natural log expression in a ParsedAux block is causing an invalid function error.  The same error does not occur when I replace the natural log with a base 10 log (a sanity check to see if other mathematical functions could be used here).  Could you advise on whether the natural log function is supported in the ParsedAux auxkernel?  Many thanks again for the help with this.

**Reply by GiudGiud** [reply]
According to the documentation for the parser, it should be supported.
Are you writing it ln() or log()? Should be the latter
http://warp.povusers.org/FunctionParser/fparser.html

**Reply by LukeFisherUKAEA** [reply]
Ah, that's the issue.  I was using the former.  Thanks for the link to the documentation.
